website/content/specification/syntax/constraints.md

```diff
@@ -14,27 +14,79 @@ TODO: P3: Address issue https://github.com/usnistgov/metaschema/issues/325
+
+The following subsections detail the processing requirements for the `@extension` and `@allow-other` attributes.
+
+### `@extension`
+
+For each `<allowed-values>` constraint, the `@extension` attribute MUST be one of the following values.
+
+- **`none`:** There can be no other matching `<allowed-values>` constraint for the same target value. This is the least permissive option.
+
+- **`model`:** (default) Multiple matching `<allowed-values>` constraints are allowed for the same target value as long as the constraints are defined in the same model. Constraints sourced from outside the model are not allowed.
+
+    All allowed-values constraints declared within a Metaschema model matching the same @target can be combined. If a matching constraint within the model has allow-other="no", then constraints declared externally from the model are not allowed. **This is the implicit default value if no `@extension` is provided.**
+
+- **`external`:** Multiple matching `<allowed-values>` constraints are allowed for the same target value, which can be sourced from within the model or externally through a set of external constraints.
+
+    All allowed-values constraints, declared within the model and externally through an extension, that match the same @target can be combined. This is the most permissive option.
+
+One of the following requirements MUST apply when processing a value's *target set* to validate it.
+
+1. The *target set* MUST contain a single `<allowed-values>` constraint with the `@extension` attribute value `none`.
+1. All `<allowed-values>` constraints in the *target set* MUST have the `@extension` attribute value `model` and originate from a *model* source.
+1. All `<allowed-values>` constraints in the *target set* MUST have the `@extension` attribute value `external` and originate from either a *model* or *extension* source.
+1. An error MUST be raised indicating the *target set* is invalid.
+
+### `@allow-other`
+
+The *expected value set* can be considered *open* or *closed*.
+
+- **open:** In an open set, the actual value can be any value. The *expected value set* provides suggested values.
+- **closed:** In a closed set, the actual value is expected to match a value in the *expected value set*.
+
+For each `<allowed-values>` constraint, the `@allow-other` attribute MUST be one of the following values.
+
+- **`yes`:** Identifies the *expected value set* as *open*, as long as no other `<allowed-values>` constraint in the *target set* has `@allow-other="no"` declared explicitly or implicitly.
+- **`no`:** (default) Identifies the *expected value set* as *closed*. **This is the implicit default value if no `@allow-other` is provided.**
+
+One of the following requirements MUST apply when processing the a value's *target set* to determine the *expected value set*.
+
+1. One `<allowed-values>` constraint in the *target set* MUST have the `@allow-other` attribute value `no`. The *expected value set* is *closed*.
```

Comment:
Effectively this means that any metaschema author can close a set of all allowed-values over a given value, by defining an `allowed-values` with `allow-other='no'`. This concerns me a little. For example, a metaschema has:

```
<allowed-values target=".//prop[@name='boo']" allow-other='yes'>... </allowed-values>
```

then an author introduces a new constraint:

```
<allowed-values target=".//part[@name='bah']//prop[@name='boo']" allow-other='no'>... </allowed-values>
```

isn't the effect that the new constraint turns the old one into 'closed' in some places and 'open' (as originally intended), in others?

Is this what is wanted? I am not sure the answer is not 'yes'. But it may be harder to explain and rationalize than a simpler 'gigo' rule that forces metaschema authors to be more careful (by running two checks instead of one and exposing rather than reconciling conflicts and contradictions).

This is not a blocker but maybe something that requires further discussion, documentation and field testing.

Reply:
@wendellpiez Some comments below.

> Effectively this means that any metaschema author can close a set of all allowed-values over a given value, by defining an `allowed-values` with `allow-other='no'`. This concerns me a little. For example, a metaschema has:
> 
> ```
> <allowed-values target=".//prop[@name='boo']" allow-other='yes'>... </allowed-values>
> ```
> 
> then an author introduces a new constraint:
> 
> ```
> <allowed-values target=".//part[@name='bah']//prop[@name='boo']" allow-other='no'>... </allowed-values>
> ```
> 
> isn't the effect that the new constraint turns the old one into 'closed' in some places and 'open' (as originally intended), in others?

Yes, but only for props contained in a "bah" part. All other "boo" props are still open under this algorithm.

> Is this what is wanted? I am not sure the answer is not 'yes'.

Yes. It is desirable as a model author to have fine-grained control around when value sets are considered open or closed under specific conditions. It is not enough to just have purely open or closed value sets. The OSCAL use case has shown this.

>  But it may be harder to explain and rationalize than a simpler 'gigo' rule that forces metaschema authors to be more careful (by running two checks instead of one and exposing rather than reconciling conflicts and contradictions).

This is a matter of taste. The current deterministic solution allows the model author to rely on the convergence provided by the solution or to approach writing the content using multiple checks. This flexibility is desired. Either way, there are clear rules on how an implementation MUST behave.

> This is not a blocker but maybe something that requires further discussion, documentation and field testing.

We are providing a solution here that: 1) does not break OSCAL use, 2) is deterministic in all cases, and 3) can be tested through implementation. These are the 3 objectives we agreed to address.

To discuss alternatives we need to go beyond talking conceptually. Are you willing to write an alternative specification around this feature for discussion that addresses these 3 objectives?

Reply:
IIUC my question is less the design of the feature but rather whether the added complexity is really needed, addressing a real use case (i.e., "seen in the field") that is not better addressed some other way. I have been assured there are such use cases in OSCAL. Maybe we should start there? Partly b/c those use cases also show nice testing for them that must implement. Where are some clashing or otherwise-clashing constraints I can look at, as instances of where we want this behavior?

Reply:
We just pushed a lot of updates to discuss the ordering and processing mechanics of constraints was not there before for review. I will close this for now, feel free to reopen if those updates are not clear (I suppose it will be discussed in the community meeting coming up).